# HG changeset patch
# User David Champion <dgc@uchicago.edu>
# Date 1454287308 0
#      Mon Feb 01 00:41:48 2016 +0000
# Branch gentoo-1.7
# Node ID 8eac2e5c1219127e1bf85880dfd15370f36179f8
# Parent  d075b5c2a28c4e6bf3c97300c3799990467fd158
[PATCH] feature: nested-if
From d04d4e08a8cd28b7db140c1d5c90c0a4e75d3c42 Mon Sep 17 00:00:00 2001
Allow complex nested conditions in the index_format
---
 muttlib.c | 55 +++++++++++++++++++++++++++++++++++++++++++++++++------
 1 file changed, 49 insertions(+), 6 deletions(-)

diff -r d075b5c2a28c -r 8eac2e5c1219 PATCHES
--- a/PATCHES	Sun Jan 31 17:17:10 2016 +0000
+++ b/PATCHES	Mon Feb 01 00:41:48 2016 +0000
@@ -1,3 +1,4 @@
+patch-nested-if-neomutt
 patch-initials-neomutt
 patch-ifdef-neomutt
 patch-browser-sort-additions-neomutt
diff -r d075b5c2a28c -r 8eac2e5c1219 muttlib.c
--- a/muttlib.c	Sun Jan 31 17:17:10 2016 +0000
+++ b/muttlib.c	Mon Feb 01 00:41:48 2016 +0000
@@ -1216,6 +1216,24 @@ void mutt_FormatString (char *dest,		/* 
 
       if (*src == '?')
       {
+	/* change original %? to new %< notation */
+	/* %?x?y&z? to %<x?y&z> where y and z are nestable */
+	char *p = (char *) src;
+	*p = '<';
+	for ( ; *p && *p != '?'; p++);
+	  /* nothing */
+	if (*p == '?') {
+	  p++;
+	}
+	for ( ; *p && *p != '?'; p++);
+	  /* nothing */
+	if (*p == '?') {
+	  *p = '>';
+	}
+      }
+
+      if (*src == '<')
+      {
 	flags |= MUTT_FORMAT_OPTIONAL;
 	src++;
       }
@@ -1242,6 +1260,8 @@ void mutt_FormatString (char *dest,		/* 
 
       if (flags & MUTT_FORMAT_OPTIONAL)
       {
+	int lrbalance;
+
         if (*src != '?')
           break; /* bad format */
         src++;
@@ -1249,8 +1269,20 @@ void mutt_FormatString (char *dest,		/* 
         /* eat the `if' part of the string */
         cp = ifstring;
 	count = 0;
-        while (count < sizeof (ifstring) && *src && *src != '?' && *src != '&')
-	{
+	lrbalance = 1;
+        while ((lrbalance > 0) && (count < sizeof (ifstring)) && *src) {
+	  if (*src == '\\') {
+	    src++;
+	    *cp++ = *src++;
+	  } else if ((src[0] == '%') && (src[1] == '<')) {
+	    lrbalance++;
+	  } else if (src[0] == '>') {
+	    lrbalance--;
+	  }
+	  if (lrbalance == 0)
+	    break;
+	  if ((lrbalance == 1) && (src[0] == '&'))
+	    break;
           *cp++ = *src++;
 	  count++;
 	}
@@ -1261,9 +1293,20 @@ void mutt_FormatString (char *dest,		/* 
 	  src++; /* skip the & */
 	cp = elsestring;
 	count = 0;
-	while (count < sizeof (elsestring) && *src && *src != '?')
-	{
-	  *cp++ = *src++;
+	while ((lrbalance > 0) && (count < sizeof (elsestring)) && *src) {
+	  if (*src == '\\') {
+	    src++;
+	    *cp++ = *src++;
+	  } else if ((src[0] == '%') && (src[1] == '<')) {
+	    lrbalance++;
+	  } else if (src[0] == '>') {
+	    lrbalance--;
+	  }
+	  if (lrbalance == 0)
+	    break;
+	  if ((lrbalance == 1) && (src[0] == '&'))
+	    break;
+          *cp++ = *src++;
 	  count++;
 	}
 	*cp = 0;
@@ -1271,7 +1314,7 @@ void mutt_FormatString (char *dest,		/* 
 	if (!*src)
 	  break; /* bad format */
 
-        src++; /* move past the trailing `?' */
+        src++; /* move past the trailing `>' (formerly '?') */
       }
 
       /* handle generic cases first */
